feat: added control to list font pairs inside typography section

preda-bogdan opened the pull request:
### Summary
<!-- Please describe the changes you made. -->
Added a new section under Typography called `Font pairings`.
Added a new control for that section that holds the available Font pairings, and allows the user to change the Heading and body font at the same time.

A new filter was added `neve_font_pairings`. It can be used to change the font pairings listed.

The font pair array allows for setting preview sizes for the font that are being used when displaying the font inside the control. This helps us have control over how the fonts are being displayed and avoid issues with fonts going out of bounds.

### Will affect visual aspect of the product
<!-- It includes visual changes? -->
YES

### Screenshots <!-- if applicable -->
![image](https://user-images.githubusercontent.com/23024731/211565556-fcae0c83-093f-47fa-a538-b7a1f35bac40.png)

![image](https://user-images.githubusercontent.com/23024731/211565647-22a751c5-dbbb-4d80-a050-d4a99e93b1a8.png)

![image](https://user-images.githubusercontent.com/23024731/211565722-b061a6fc-578e-43a7-98c9-01c22e2eba69.png)

### Test instructions
<!-- Describe how this pull request can be tested. -->
1. Check that the fonts update correctly for heading and body when a pair is selected.
2. Check that they display correctly on the front-end and inside the Customizer.
3. Check that when global colors are changed the preview for the control changes. E.g. the colors used to display the preview inside the control are: `Dark Background`, `Text Dark Background` for hover and `Site Background`, `Text Color` for normal state.
4. Check the values are reset to the Default value of the typography Control when `Reset to customizer defaults`
 is clicked.
<!-- Issues that this pull request closes. -->
Closes Codeinwp/neve-pro-addon#2190.
<!-- Should look like this: `Closes #1, #2, #3.` . -->

selul commented on assets/apps/customizer-controls/src/typography-font-pair/FontPairSelector.tsx:
```diff
@@ -0,0 +1,154 @@
+					}}
+				>
+					<span className="dashicons dashicons-image-rotate"></span>
+					{__('Reset to customiser defaults', 'neve')}
```
@preda-bogdan we can re-use some of the available strings, i.e Reset Value

selul commented on inc/customizer/options/typography.php:
```diff
@@ -28,26 +28,75 @@ class Typography extends Base_Customizer {
+							'link' => apply_filters( 'neve_external_link', 'https://docs.themeisle.com/article/1340-neve-typography', esc_html__( 'Learn more', 'neve' ) ),
+						],
+					),
+					'label'       => esc_html__( 'Font pairings', 'neve' ),
```
@JohnPixle @mghenciu @abaicus @preda-bogdan how about if we change this to `Customize Fonts` to avoid adding new strings?

abaicus replied:
It's worth changing it to avoid adding the new strings.

selul commented on inc/customizer/options/typography.php:
```diff
@@ -28,26 +28,75 @@ class Typography extends Base_Customizer {
+					'input_attrs' => array(
+						'pairs'       => $pairs,
+						'description' => [
+							'text' => __( 'Choose Font family presets for your Headings and Text.', 'neve' ),
```
is this required @preda-bogdan ? How about if we just use the link?

preda-bogdan commented:
@selul I've updated the strings, I assumed we would keep them since they were part of the design. One thing here the description now looks a bit empty:

![image](https://user-images.githubusercontent.com/23024731/211575066-56a9b221-578d-4bb6-b4d9-0c22c6230f7d.png)

Maybe the guys from the design team can suggest some changes here or maybe we can drop this or try to move it inside the control header.

abaicus commented on assets/apps/customizer-controls/src/typography-font-pair/FontPairSelector.tsx:
```diff
@@ -0,0 +1,153 @@
+				? pairs.map((preset, index) => {
+						const { bodyFont, headingFont } = preset;
+						const label = `${headingFont.font} & ${bodyFont.font}`;
+						const headingStyle: FontStyle = {
+							fontFamily: `${headingFont.font}, sans-serif`,
+						};
+						const bodyStyle: FontStyle = {
+							fontFamily: `${bodyFont.font}, sans-serif`,
+						};
```
I think these can be typed as `CSSProperties`. No need to create a new type 😁.

abaicus commented on assets/apps/customizer-controls/src/typography-font-pair/FontPairSelector.tsx:
```diff
@@ -0,0 +1,153 @@
+						setSelected(-1);
+					}}
+				>
+					<span className="dashicons dashicons-image-rotate"></span>
```
Nitpicking here, but empty tags can usually be written as self-closing:
```suggestion
					<span className="dashicons dashicons-image-rotate"/>
```

mghenciu commented:
🔝 work Bogdan!

Just a small visual issue on the right-bottom side ob the Checkmark box, there looks to be some unnecessary space (visible both on Dark or Light mode): 
![Screenshot 2023-01-10 at 17 02 47](https://user-images.githubusercontent.com/52494172/211586643-9817deda-63ee-4bf7-a321-2ba3a21c2941.png)

Playing a bit more with the border radius of the ::before element, should reveal a solution.
___
Regarding the Keywords, I think John intention with that paragraph (_Choose Font family presets for your Headings and Text. Learn more._) was to be a quick explainer of what this feature is + a link to a Section in the Docs just about this.
 If we are going this way, we can ask the support team to create a short section for this.
 
 Now if terms of text, we can use something similar we have for the **Fallback fonts**, but in this case say something like:
 **_Learn more about customizing fonts_**, hoping that Customizing is already translated somewhere.
 
![Screenshot 2023-01-10 at 17 15 44](https://user-images.githubusercontent.com/52494172/211589543-814f592b-dab9-4243-acfc-c0b01474c7b5.png)

___
**But we can also drop this (for the initial phase),** because at this point this feature is not very complex to need a explainer, meaning that users can just browse through those presents and see directly how they work, so there's not a lot to explain imo. 

cc. @JohnPixle

abaicus commented on assets/apps/customizer-controls/src/builder/scss/_typography-font-pair-section.scss:
```diff
@@ -0,0 +1,81 @@
+			height: 27px;
+			background-color: var(--wp-admin-theme-color);
+			color: #fff;
+			border-radius: 6px 0 4px 0;
```
I think this can be 6px 0 0; to avoid the small visual issue seen in the screenshot attached to the initial PR body: 
![image](https://user-images.githubusercontent.com/15010186/211591228-abd21ad2-b80c-4b98-836e-623af580ccb8.png)

abaicus reviewed (changes requested):
@preda-bogdan Things look good. 👍🏻 Left some small stuff below 🚀

irinelenache commented:
@preda-bogdan Nice job 🚀 Tested and here is what i found so far:

- [ ] The spacing is not the same for all buttons because sometimes the heading text goes on two lines https://vertis.d.pr/i/7CU2d3 (Probably this happens because i use a 14inch screen)
- [ ] If i have some fonts previously selected for Headings and Body, after selecting and deselecting a preset, the Body font is lost in customizer https://vertis.d.pr/v/1j5u7h In frontend, both Body and Headings fonts are lost
- [ ] One time the selected icon looked [like this](https://vertis.d.pr/i/inNMVB) for me. I refreshed the customizer and it worked fine after, i didn't do anything special when it came up. I'll try to see if i can replicate it again
- [ ] The new tab icon [here](https://vertis.d.pr/i/nTXty8) should be after the **Learn more** text

Besides these, everything's fine on my end 🙏 

Issues can be checked here:
```
https://font-pairing.vertisite.cloud/wp-admin
Username: woobooster
Password: Js4Oa61d
```

preda-bogdan commented:
@irinelenache 
Updated the new tab icon position:
![image](https://user-images.githubusercontent.com/23024731/212909013-9a044c29-f2b4-4048-b1b9-2263c0eb056f.png)

The selected tab icon might have been an issue with the resource not loaded but I can not replicate it locally. Since we are using the icons from WordPress I don't think it's an issue, but do let me know if you find a way to replicate this.

I fixed the spacing and the font selection on reset. It should work also if you don't publish the changes (of course only inside the Customizer, for the front end you still need to publish.)

@JohnPixle Thank you for the suggestion, I would love to use the clamp function however the support is not yet there, especially on Safari. However, I did something similar, if the font exceeds a specific size we just hide the overflow and replace it with the ellipsis (see image below).
![image](https://user-images.githubusercontent.com/23024731/212910143-640ed891-e6e6-4095-8729-95d224b9cb6e.png)

However, for the font sizes we use now, this does not display if you force them to one line it is enough to display correctly:
![image](https://user-images.githubusercontent.com/23024731/212910380-026861ea-c82d-47e6-9916-a4aa69a1012c.png)

I clamped the body font at 24px and the heading at 27px we can adjust if needed.

preda-bogdan commented:
@irinelenache I can't replicate this locally either in Chrome or Firefox.
Can you confirm it is not a caching issue?
I use the neve-pro development version, what version of neve-pro are you using in this instance?

Thank you!

irinelenache commented:
@preda-bogdan Here are the setup details:

```
Chrome Version 109.0.5414.87
MacBook Pro (14-inch, 2021)
macOS Monterey Version 12.2.1 (21D62)
```

It also happens with Safari, so it might be again because of the screen size

preda-bogdan commented:
@irinelenache I think I solved the issue that you experienced can you please confirm when you have some time?
Thank you!

irinelenache commented:
@preda-bogdan I checked again and it is the same on my screen, but it works fine on browserstack 👍 

I think we can move this to **Ready to merge** and have more people checking it with the final Neve 3.5 build ✅

cristian-ungureanu commented:
Hey @preda-bogdan, you've made the PR to development. I changed it to new/3-5, and now there are some conflicts. Can you have a look and resolve the conflicts, please?

preda-bogdan commented:
@cristian-ungureanu should be fine now.
